ospsuite-r-specifics/r-code-structure.md

```diff
@@ -0,0 +1,250 @@
+#' @export
+hasDimension <- function(dimension) {
+  validateIsString(dimension)
+  dimensionTask <- .getNetTaskFromCache("DimensionTask")
+  rClr::clrCall(dimensionTask, "HasDimension", enc2utf8(dimension))
+}
+```
+
+
+As you can see in order to retrieve the Dimension Task, we call the internal function `.getNetTaskFromCache`. In order to avoid having to get them from .NET all the time, we cache them on the R side. As you can see in [get-net-task.R](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/develop/R/get-net-task.R):
+
+```
+.
+.
+.
+
+#' @title .getNetTaskFromCache
+#' @description Get an instance of the specified `.NET` Task that is retrieved
+#' from cache if already initiated. Otherwise a new task will be initiated and
+#' cached in the `tasksEnv`.
+#'
+#' @param taskName The name of the task to retrieve (**without** `Get` prefix).
+#'
+#' @return returns an instance of of the specified `.NET` task.
+#'
+#' @keywords internal
+.getNetTaskFromCache <- function(taskName) {
+  if (is.null(tasksEnv[[taskName]])) {
+    tasksEnv[[taskName]] <- .getNetTask(taskName)
+  }
+  return(tasksEnv[[taskName]])
+}
+```
+
+we cache the tasks in the `tasksEnv[]` list. If we do not find a task in the cache we retrieve it from .NET through an rClr call and we also add it to the cache for future use.
+
+### Tests
+
+The OSPSuite-R package is well tested and you can find all the code for the tests as usual under [testthat](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/tree/develop/tests/testthat). Apart from guaranteeing the correct and consistent functioning of the package, the tests are also a good entry point to finding out about how the rest of the code works, f.e. how objects get created and used and so on.
+
+## Updating Core dlls
+
+The R package keeps local copies of the necessary dlls coming from OSPSuite.Core and PK-Sim that are necessary for it to function. When a newer version of the .NET codebase is available, those dlls need to be updated semi-manually. Those dlls have to exist under [OSPSuite-R/inst/lib/](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/tree/develop/inst/lib). 
+
+The first step to updating to a newer version of Core is updating the nuget package versions to the correct version. For this you have to manually edit the version numbers in [packages.config](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/develop/packages.config), like the version number in OSPSuite.Core seen here underneath:
+
+```
+<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  .
+  .
+  .
+
+  <package id="OSPSuite.Core" version="12.0.242"/>
+  .
+  .
+
+</packages>
+```
+
+Then you have to go to the command line in the OSPSuite-R repository main folder and run:
+```
+nuget restore packages.config -PackagesDirectory packages
+```
+Note that nuget has to have been added to the Path, otherwise you will need to provide the full path to nuget.exe.
```

Comment:
Add a link to some web resource on how to add something to the PATH variable under windows.

Reply:
done